File: app/src/main/java/br/com/popularmovies/moviedetail/trailers/viewmodel/factories/MovieTrailerFactory.java
package br.com.popularmovies.moviedetail.trailers.viewmodel.factories;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import br.com.popularmovies.moviedetail.trailers.viewmodel.MovieTrailerViewModel;
import br.com.popularmovies.services.movieService.MovieRepositoryImpl;

public class MovieTrailerFactory implements ViewModelProvider.Factory {
    private final int movieId;
    private final MovieRepositoryImpl mMovieRepository;

    public MovieTrailerFactory(MovieRepositoryImpl mMovieRepository, int movieId) {
        this.mMovieRepository = mMovieRepository;
        this.movieId = movieId;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new MovieTrailerViewModel(mMovieRepository, this.movieId);
    }
}

File: app/src/main/java/br/com/popularmovies/moviedetail/trailers/viewmodel/factories/MovieTrailerFactory.kt
package br.com.popularmovies.moviedetail.trailers.viewmodel.factories

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import br.com.popularmovies.moviedetail.trailers.viewmodel.MovieTrailerViewModel
import br.com.popularmovies.repositories.movie.MovieRepository
import javax.inject.Inject

class MovieTrailerFactory @Inject constructor(
        private val mMovieRepository: MovieRepository,
        private val movieId: Int
) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return MovieTrailerViewModel(mMovieRepository, movieId) as T
    }
}